zebrad/tests/acceptance.rs

```diff
@@ -860,6 +864,10 @@ fn create_cached_database_height(network: Network, height: Height) -> Result<()>
     let network = format!("network: {},", network);
     child.expect_stdout_line_matches(&network)?;
     child.expect_stdout_line_matches(STOP_AT_HEIGHT_REGEX)?;
+
+    child.expect_stdout_line_matches("starting legacy chain check")?;
+    child.expect_stdout_line_matches("no legacy chain found")?;
```

Comment:
Test coverage: Can you also add these checks to `sync_until`?

Those tests are enabled in CI, so we can check we got the log order right.

Reply:
I tried to do this however the output if the child there stops before the check is done:

```
Jun 24 21:02:39.528  INFO {zebrad="bee2cca" net="Main"}: zebrad::commands::start: Starting zebrad
Jun 24 21:02:39.528  INFO {zebrad="bee2cca" net="Main"}: zebrad::commands::start: config=ZebradConfig { consensus: Config { checkpoint_sync: false }, metrics: MetricsSection { endpoint_addr: None }, network: Config { listen_addr: 127.0.0.1:0, network: Mainnet, initial_mainnet_peers: {"dnsseed.z.cash:8233", "mainnet.is.yolo.money:8233", "mainnet.seeder.zfnd.org:8233", "dnsseed.str4d.xyz:8233"}, initial_testnet_peers: {"testnet.seeder.zfnd.org:18233", "testnet.is.yolo.money:18233", "dnsseed.testnet.z.cash:18233"}, peerset_initial_target_size: 50, crawl_new_peer_interval: 30s }, state: Config { cache_dir: "/tmp/zebrad_tests.ShIQzAXkoZkZ/state", ephemeral: false, debug_stop_at_height: Some(0) }, tracing: TracingSection { use_color: true, filter: None, endpoint_addr: None, flamegraph: None, use_journald: false }, sync: SyncSection { max_concurrent_block_requests: 50, lookahead_limit: 2000 } }
Jun 24 21:02:39.528  INFO {zebrad="bee2cca" net="Main"}: zebrad::commands::start: initializing node state
Jun 24 21:02:39.529  INFO {zebrad="bee2cca" net="Main"}: zebra_state::config: the open file limit is at or above the specified limit new_limit=1024 current_limit=1024 hard_rlimit=Some(4096)
Jun 24 21:02:39.569  INFO {zebrad="bee2cca" net="Main"}: zebra_state::service::finalized_state: Opened Zebra state cache at /tmp/zebrad_tests.ShIQzAXkoZkZ/state/state/v5/mainnet
Jun 24 21:02:39.570  INFO {zebrad="bee2cca" net="Main"}: zebra_state::service::finalized_state: state is already at the configured height debug_stop_at_height=Height(0) tip_height=Height(0) tip_hash=block::Hash("00040fe8ec8471911baa1db1266ea15dd06b4a8a5c453883c000b031973dce08")
```

Reply:
You could try something like:
```rust
if child.expect_stdout_line_matches("starting legacy chain check").is_ok() {
    child.expect_stdout_line_matches("no legacy chain found")?;
}
```

That way, if the legacy check is skipped, the test still passes.

(It's ok if the legacy chain check is skipped, because Zebra still works. But if it fails and panics on a good chain, we want to catch that in the tests.)

Reply:
That will not work either:

```
The application panicked (crashed).
Message:  child must capture stdout to call expect_stdout_line_matches
```

Reply:
### Priority

This test is pretty important, because we don't have any other way of testing the legacy chain check against cached state. But we can't do it until we add an NU5 activation height.

So let's open another ticket for this change?

Since we'll be relying on the proptests to merge this PR, we'll have to make sure they're testing the right height ranges and data.

### Details

That's a confusing panic message, see PR #2406 for a better one.

What's happening is that the output gets dropped and the process killed when there isn't any match. So my original idea wouldn't have worked, because:
`if child.expect_stdout_line_matches("starting legacy chain check")`
would consume all the output and kill the process.

Instead, we'll need to pass a parameter to `sync_until`, which tells it if it should check for legacy chain output.

If we expect the state to sync some blocks, and there is an NU5 activation height, we should check for legacy chain output. But if we expect to be at the stop height, or there is no NU5 activation height, we should skip the legacy chain output check, because Zebra will exit before it happens.

Reply:
Fixed by a645a74